osu.Game/Graphics/UserInterface/SegmentedGraph.cs

```diff
@@ -0,0 +1,320 @@
+            private readonly SegmentInfo?[] pendingSegments;
+
+            public SegmentManager(int tierCount)
+            {
+                pendingSegments = new SegmentInfo?[tierCount];
+            }
+
+            public void StartSegment(int tier, float start)
+            {
+                if (pendingSegments[tier] != null)
+                    throw new InvalidOperationException($"Another {nameof(SegmentInfo)} of tier {tier.ToString()} has already been started.");
+
+                pendingSegments[tier] = new SegmentInfo
+                {
+                    Tier = tier,
+                    Start = Math.Clamp(start, 0, 1)
+                };
+            }
+
+            public void EndSegment(int tier, float end)
+            {
+                SegmentInfo? pendingSegment = pendingSegments[tier];
+                if (pendingSegment == null)
+                    throw new InvalidOperationException($"Cannot end {nameof(SegmentInfo)} of tier {tier.ToString()} that has not been started.");
+
+                SegmentInfo segment = pendingSegment.Value;
+                segment.End = Math.Clamp(end, 0, 1);
+                segments.Add(segment);
+                pendingSegments[tier] = null;
+            }
+
+            public void EndAllPendingSegments()
+            {
+                foreach (SegmentInfo? pendingSegment in pendingSegments)
+                {
+                    if (pendingSegment != null)
+                    {
+                        SegmentInfo finalizedSegment = pendingSegment.Value;
+                        finalizedSegment.End = 1;
+                        segments.Add(finalizedSegment);
+                    }
+                }
+            }
+
+            public void Sort() =>
+                segments.Sort((a, b) =>
+                    a.Tier != b.Tier
+                        ? a.Tier.CompareTo(b.Tier)
+                        : a.Start.CompareTo(b.Start));
+
+            public void Add(SegmentInfo segment) => segments.Add(segment);
+
+            public void Clear()
+            {
+                segments.Clear();
+
+                for (int i = 0; i < pendingSegments.Length; i++)
+                    pendingSegments[i] = null;
+            }
+
+            public int Count => segments.Count;
+
+            public void Add(int tier, float start, float end)
+            {
+                SegmentInfo segment = new SegmentInfo
+                {
+                    Tier = tier,
+                    Start = Math.Clamp(start, 0, 1),
+                    End = Math.Clamp(end, 0, 1)
+                };
+
+                if (segment.Start > segment.End)
+                    throw new InvalidOperationException("Segment start cannot be after segment end.");
+
+                Add(segment);
+            }
+
+            public bool IsTierStarted(int tier)
+            {
+                if (tier < 0)
+                    return false;
+
+                return pendingSegments[tier].HasValue;
+            }
```

Comment:
Nit, can be formated as  ```public bool IsTierStarted(int tier) =>
                tier >= 0 && pendingSegments[tier].HasValue;``` seeing as expression bodied is generally preferred and youve used them where applicable in the file elsewhere.

Reply:
Oops, should have caught that earlier